File: arduino/ArduComm/ArduParser.cpp
/* Module to parse and serialize the data from/into the bytes received/to send
*
*  Note: Works only on version 1.0 and above of Arduino IDE. (only arduino.h)
*
*  Author:  Francisco Miguel Moreno
*  Email:   butakus@gmail.com
*  Version: 0.2
*  Date:    April 2017
*/

#include "ArduParser.h"


char parse_char(uint8_t* data)
{
    return (char) *data;
}

uint8_t parse_string(uint8_t* data, char* out_buffer)
{
    int i = 0;
    while (data[i] != 0)
    {
        out_buffer[i] = (char) data[i];
        ++i;
    }
    out_buffer[i] = 0;
    return i + 1;
}


int8_t parse_int8(uint8_t* data)
{
    return (uint8_t) *data;
}

uint16_t parse_uint16(uint8_t* data)
{
    return *(uint16_t*) data;
}

int16_t parse_int16(uint8_t* data)
{
    return *(int16_t*) data;
}

uint32_t parse_uint32(uint8_t* data)
{
    return *(uint32_t*) data;
}

int32_t parse_int32(uint8_t* data)
{
    return *(int32_t*) data;
}


float parse_float(uint8_t* data)
{
    return *(float*) data;
}


void serialize_char(char& data, uint8_t* buffer)
{
    buffer[0] = (uint8_t)data;
}

void serialize_string(char* data, uint8_t* buffer)
{
    int i = 0;
    while (data[i] != 0)
    {
        buffer[i] = (uint8_t)data[i];
        ++i;
    }
    buffer[i] = 0;
}


void serialize_int8(int8_t& data, uint8_t* buffer)
{
    buffer[0] = (uint8_t)data;
}

void serialize_uint16(uint16_t& data, uint8_t* buffer)
{
    memcpy(buffer, (unsigned char*)(&data), 2);
}

void serialize_int16(int16_t& data, uint8_t* buffer)
{
    memcpy(buffer, (unsigned char*)(&data), 2);
}

void serialize_uint32(uint32_t& data, uint8_t* buffer)
{
    memcpy(buffer, (unsigned char*)(&data), 4);
}

void serialize_int32(int32_t& data, uint8_t* buffer)
{
    memcpy(buffer, (unsigned char*)(&data), 4);
}


void serialize_float(float& data, uint8_t* buffer)
{
    memcpy(buffer, (unsigned char*)(&data), 4);
}

File: host/arduparser.py
""" Module to parse the bytes read with ArduComm into common types """

import struct


""" Parsers: byte list to common types: """


def parse_char(data):
    """ Convert the given 1-size byte list into a char """
    return chr(data[0])

def parse_str(data):
    """ Convert the given byte list into a string (char array)
        The string data must be ended in \0
    """
    return data.decode('utf8').rstrip('\x00')


def parse_int8(data, big_endian=False):
    """ Convert 1-size byte list into a 8 bit signed int """
    byte_format = '>b' if big_endian else '<b'
    return struct.unpack(byte_format, bytes(data))[0]

def parse_int16(data, big_endian=False):
    """ Convert 2 bytes into a 16 bit int """
    byte_format = '>h' if big_endian else '<h'
    return struct.unpack(byte_format, bytes(data))[0]

def parse_uint16(data, big_endian=False):
    """ Convert 2 bytes into an unsigned 16 bit int """
    byte_format = '>H' if big_endian else '<H'
    return struct.unpack(byte_format, bytes(data))[0]

def parse_int32(data, big_endian=False):
    """ Convert 4 bytes into a 32 bit int """
    byte_format = '>i' if big_endian else '<i'
    return struct.unpack(byte_format, bytes(data))[0]

def parse_uint32(data, big_endian=False):
    """ Convert 4 bytes into an unsigned 32 bit int """
    byte_format = '>I' if big_endian else '<I'
    return struct.unpack(byte_format, bytes(data))[0]

def parse_float(data, big_endian=False):
    """ Convert 4 bytes into a 32 bit float """
    byte_format = '>f' if big_endian else '<f'
    return struct.unpack(byte_format, bytes(data))[0]


""" Serializers: common types to byte list: """


def serialize_char(data):
    """ Convert the given char into a bytes object """
    return bytes(data, 'utf8')

def serialize_str(data):
    """ Convert the given string into a bytes object """
    # Sanitize the input first
    if data[-1] != '\x00':
        data += '\x00'
    return bytes(data, 'utf8')

def serialize_int8(data, big_endian=False):
    """ Convert a 8 bit signed int into a bytes object """
    byte_format = '>b' if big_endian else '<b'
    return struct.pack(byte_format, data)

def serialize_int16(data, big_endian=False):
    """ Convert a 16 bit int into 2 bytes """
    byte_format = '>h' if big_endian else '<h'
    return struct.pack(byte_format, data)

def serialize_uint16(data, big_endian=False):
    """ Convert an unsigned 16 bit int into 2 bytes """
    byte_format = '>H' if big_endian else '<H'
    return struct.pack(byte_format, data)

def serialize_int32(data, big_endian=False):
    """ Convert a 32 bit int into 4 bytes """
    byte_format = '>i' if big_endian else '<i'
    return struct.pack(byte_format, data)

def serialize_uint32(data, big_endian=False):
    """ Convert an unsigned 32 bit int into 4 bytes """
    byte_format = '>I' if big_endian else '<I'
    return struct.pack(byte_format, data)

def serialize_float(data, big_endian=False):
    """ Convert a 32 bit float into 4 bytes """
    byte_format = '>f' if big_endian else '<f'
    return struct.pack(byte_format, data)
